Add classifier-spdx fuzz target for afl++

dextero opened the pull request:
The new target is an executable (`fuzz-classifier-spdx`) that reads all data available on stdin and passes it to both strict and lenient SPDX licence classifiers.

Usage:
* Install afl++ (https://github.com/AFLplusplus/AFLplusplus)
* Build & start fuzzing with `make fuzz` (or `make fuzz-coverage` for fuzzing with coverage data, see below). This starts fuzzing based on all the input files from `test/fuzz/input/*`, and can be stopped with Ctrl+C.
* In case of any crashes, the `test/fuzz/output/default/crashes` directory will contain inputs that cause `fuzz-classifier-spdx` to crash. They can be used to reproduce issues.
* (optional) Use https://github.com/mrash/afl-cov to view code coverage of the fuzzing process:

```
afl-cov --overwrite -d ./test/fuzz/output \
        --coverage-cmd "cat AFL_FILE | $PWD/build/fuzz-classifier-spdx" \
        --code-dir ./
xdg-open ./test/fuzz/output/cov/web/index.html
```

I only tried this on Ubuntu, and had to provide my own implementation of `rpmvercmp` (not included here) for it to compile because it apparently is missing from the librpm I got from the repo. Please double check if it works well for you.

dextero committed:
Add classifier-spdx fuzz target for afl++

The new target is an executable (`fuzz-classifier-spdx`) that reads all
data available on stdin and passes it to both strict and lenient SPDX
licence classifiers.

Usage:
* Install afl++ (https://github.com/AFLplusplus/AFLplusplus)
* Build & start fuzzing with `make fuzz` (or `make fuzz-coverage` for
  fuzzing with coverage data, see below). This starts fuzzing based on
  all the input files from `test/fuzz/input/*`, and can be stopped with
  Ctrl+C.
* In case of any crashes, the `test/fuzz/output/default/crashes`
  directory will contain inputs that cause `fuzz-classifier-spdx` to
  crash. They can be used to reproduce issues.
* (optional) Use https://github.com/mrash/afl-cov to view code coverage
  of the fuzzing process:

```
afl-cov --overwrite -d ./test/fuzz/output \
        --coverage-cmd "cat AFL_FILE | $PWD/build/fuzz-classifier-spdx" \
        --code-dir ./
xdg-open ./test/fuzz/output/cov/web/index.html
```

I only tried this on Ubuntu, and had to provide my own implementation of
`rpmvercmp` (not included here) for it to compile because it apparently
is missing from the librpm I got from the repo. Please double check if
it works well for you.

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

suve commented on Makefile:
```diff
@@ -84,6 +84,15 @@ remove: install/prepare
+
+.PHONY: fuzz-coverage
+fuzz-coverage: CFLAGS += --coverage
+fuzz-coverage: LDLIBS += -lgcov
```
In other parts of the Makefile, we use `LDFLAGS` for holding `-llibrary` flags. Please either use `LDFLAGS` here or rename the current variable to `LDLIBS`.

dextero replied:
I ran into issues when `-lfoo` flags were defined before `.o` - in some cases, the linker may ignore libraries passed at command line, if no object files preceding it make use of them (see e.g. https://wiki.gentoo.org/wiki/Project:Quality_Assurance/As-needed#Importance_of_linking_order)

suve commented on test/fuzz/classifier-spdx.c:
```diff
@@ -0,0 +1,76 @@
+
+const size_t CHUNK_SIZE = 4096;
+
+struct buffer {
```
This looks like a buffer that will re-allocate memory so that whatever data was passed into it will be stored in a single chunk. This functionality is already provided by `struct ReBuffer` found in `src/buffers.h`.

dextero replied:
Nice, thanks!

suve commented on test/fuzz/classifier-spdx.c:
```diff
@@ -0,0 +1,76 @@
+	return buf.data;
+}
+
+int main(int argc, char* argv[]) {
```
I feel this could be more useful if it was modified slightly to test only one classifier at a time, and take it as an argument - i.e. take a look at `argv[1]` and based on that use the proper `->*Classifier`. We could then use Make targets for convenience, e.g. introduce a `fuzz-classifier-spdx-strict` target that'd call this program with the proper arguments.

dextero replied:
On the other hand, it's easier to evaluate coverage completeness of the fuzzing process - the lenient/strict distinction is just a handful of if statements.

Anyway, done, with loose classifier support, because why not.

suve commented on Makefile:
```diff
@@ -84,6 +84,15 @@ remove: install/prepare
 test: build/test-suite
 	./build/test-suite
 
+.PHONY: fuzz
```
1. Please add these to the `.PHONY` declaration at the top.

2. Please add a note about this to the `help` target.

dextero replied:
Done

suve commented on test/fuzz/classifier-spdx.c:
```diff
@@ -0,0 +1,76 @@
+#include "../licences.h"
```
Please add the licensing header/comment here. You can also update the year next to your name in the `Makefile`.

dextero replied:
Done

suve reviewed (changes requested):
First of all - big thanks for looking into this! I'll try running this some time later today. In the meantime, a few quick comments.

> I only tried this on Ubuntu, and had to provide my own implementation of rpmvercmp (not included here) for it to compile because it apparently is missing from the librpm I got from the repo.

This is because `rpmvercmp()` comes from `librpm`. The main program can be compiled with out without this dependency by passing `WITH_LIBRPM=0` to Make.

dextero committed:
Add licence headers

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

dextero committed:
Use LDLIBS, not LDFLAGS for libraries passed to the linker

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

dextero committed:
Clean up fuzz targets in Makefile, add help

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

dextero committed:
ReBuffer: remove double pointer

This makes it slightly easier to extract the buffer out of the struct.

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

dextero committed:
Fuzz test: use ReBuffer

dextero committed:
Classifier fuzz test: pick classifier based on cmdline arg

Also:
* Add a possibility to fuzz loose classifier, because at this point
  might as well
* Rename from fuzz-classifier-spdx to fuzz-classifier

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

dextero committed:
classifier-spdx.c: initialize plusSign

There's no way of it being used uninitialized, but afl-gcc *really*
can't figure that out, and I couldn't find a way to disable the warning
being a hard error without throwing out all other warnings.

Signed-off-by: Marcin Radomski <marcin@mradomski.pl>

suve commented:
Thanks for updating the pull request. Looks good to me (though I'll probably still tweak some things).

Once again, big thanks for looking into this and figuring it out!
